File: packages/logic/src/mechanics/updateAllTick.ts
// Per-tick accelerator-state aggregator. Lifted from
// packages/web_ui/src/Synergism.ts (updateAllTick).
//
// Computes the full accelerator stack for the current tick:
//   - totalAccelerator / costDivisor (bookkeeping)
//   - freeUpgradeAccelerator (intermediate snapshot used elsewhere)
//   - freeAccelerator (from upgrades + boosts + runes + cubes + multipliers
//     + hepteracts + corruptions)
//   - acceleratorPower / acceleratorEffect / acceleratorEffectDisplay
//   - generatorPower
//
// All "effect" inputs (achievement rewards, rune effects, cube blessings,
// hepteract effects, viscosity power, challenge 15 reward) are pre-evaluated
// by web_ui and passed in. G.acceleratorMultiplier is also pre-evaluated —
// the web_ui shim calls calculateAcceleratorMultiplier() before invoking
// updateAllTick so this slot is fresh.

import { Decimal } from '../math/bignum'
import { CalcECC } from './challenges'

export interface UpdateAllTickInput {
  // ─── Direct player state ──────────────────────────────────────────────

  /** player.acceleratorBought — base for totalAccelerator. */
  acceleratorBought: number
  /** player.multiplierBought — feeds upgrade-8's `floor(mult/7)`. */
  multiplierBought: number
  /** player.upgrades[8] — non-zero enables the multiplier-bought-derived bonus. */
  upgrade8: number
  /** player.upgrades[11] — when > 0.5 and r-chal ≠ 7, enables generatorPower formula. */
  upgrade11: number
  /** player.upgrades[21] — non-zero adds +5. */
  upgrade21: number
  /** player.upgrades[22] — non-zero adds +4. */
  upgrade22: number
  /** player.upgrades[23] — non-zero adds +3. */
  upgrade23: number
  /** player.upgrades[24] — non-zero adds +2. */
  upgrade24: number
  /** player.upgrades[25] — non-zero adds +1. */
  upgrade25: number
  /** player.upgrades[32] — non-zero adds min(500, floor(log1e25(prestigePoints+1))). */
  upgrade32: number
  /** player.upgrades[45] — non-zero adds min(2500, floor(log10(transcendShards+1))). */
  upgrade45: number
  /** player.upgrades[46] — when > 0.5, sets tuSevenMulti to 1.05 (else 1). */
  upgrade46: number
  /** player.prestigePoints — log-base argument for upgrade-32. */
  prestigePoints: Decimal
  /** player.transcendShards — log-base argument for upgrade-45. */
  transcendShards: Decimal
  /** player.challengecompletions[1] — used in t-chal 1 acceleratorPower formula. */
  c1Completions: number
  /** player.challengecompletions[2] — fed through CalcECC by caller (transcend) for ecc2tr. */
  c2Completions: number
  /** player.challengecompletions[7] — fed through CalcECC by caller (reincarnation) for ecc7r. */
  c7Completions: number
  /** player.currentChallenge.transcension — controls acceleratorPower overrides for 1/2/3. */
  transcensionChallenge: number
  /** player.currentChallenge.reincarnation — controls acceleratorPower / acceleratorEffect overrides for 7/10. */
  reincarnationChallenge: number
  /** player.researches[18] — adds +2 per level to the totalAcceleratorBoost factor. */
  research18: number
  /** player.researches[19] — adds +2 per level to the totalAcceleratorBoost factor. */
  research19: number
  /** player.researches[20] — adds +3 per level to the totalAcceleratorBoost factor. */
  research20: number
  /** player.platonicUpgrades[6] — contributes (1 + n/30) to the exponent on `a`. */
  platonicUpgrade6: number
  /** player.unlocks.prestige — gates the multiplicative-accelerators rune effect. */
  prestigeUnlocked: boolean
  /** player.corruptions.used.viscosity — sources viscosityPower and triggers >=15 / >=16 cuts. */
  viscosityCorruptionLevel: number

  // ─── Pre-evaluated effects / rewards / blessings ──────────────────────

  /** `+getAchievementReward('accelerators')` — added to base `a` (after upgrades). */
  acceleratorsAchievement: number
  /** `+getAchievementReward('acceleratorPower')` — additive term inside acceleratorPower. */
  acceleratorPowerAchievement: number
  /** `getRuneEffects('speed', 'multiplicativeAccelerators')` — multiplies `a` when prestige unlocked. */
  multiplicativeAcceleratorsRune: number
  /** `getRuneEffects('speed', 'acceleratorPower')` — additive term inside acceleratorPower. */
  acceleratorPowerRune: number
  /** `calculateAcceleratorCubeBlessing()` — adds to the totalAcceleratorBoost factor. */
  acceleratorCubeBlessing: number
  /** `getHepteractEffects('accelerator').accelerators` — added to `a` after rune+multiplier+power steps. */
  hepteractAccelerators: number
  /** `getHepteractEffects('accelerator').acceleratorMultiplier` — multiplies `a` after the additive step. */
  hepteractAcceleratorMult: number

  // ─── G inputs (pre-extracted by web_ui) ───────────────────────────────

  /** G.totalAcceleratorBoost (set by calculateTotalAcceleratorBoost prior to this call). */
  totalAcceleratorBoost: number
  /** G.acceleratorMultiplier (set by calculateAcceleratorMultiplier prior to this call). */
  acceleratorMultiplier: number
  /** G.viscosityPower[player.corruptions.used.viscosity] — viscosity-corruption exponent factor. */
  viscosityPower: number
  /** G.challenge15Rewards.accelerator.value — multiplied into `a` after the hepteract additive. */
  challenge15RewardAccelerator: number
}

export interface UpdateAllTickResult {
  /** G.totalAccelerator (= acceleratorBought + freeAccelerator). */
  totalAccelerator: number
  /** G.costDivisor — always set to 1 here, surfaced for parity. */
  costDivisor: number
  /** G.freeUpgradeAccelerator — `a` snapshot right after upgrades/achievements/ECC are folded in. */
  freeUpgradeAccelerator: number
  /** G.freeAccelerator — final floored `a` after all multiplicative + corruption gating. */
  freeAccelerator: number
  /** G.tuSevenMulti — 1.05 when upgrade46 > 0.5, else 1. */
  tuSevenMulti: number
  /** G.acceleratorPower — used as exponent base for acceleratorEffect. */
  acceleratorPower: number
  /** G.acceleratorEffect — Decimal pow result; reads totalAccelerator (+ totalMultiplier when t-chal 1). */
  acceleratorEffect: Decimal
  /** G.acceleratorEffectDisplay — `acceleratorPower * 100 - 100` wrapped in Decimal. */
  acceleratorEffectDisplay: Decimal
  /** G.generatorPower — `1.02 ^ totalAccelerator` when upgrade11 active outside r-chal 7; 1 otherwise. */
  generatorPower: Decimal
}

/**
 * Per-tick accelerator-state aggregator. See file-level comment for the
 * computation sequence; mirrors the legacy updateAllTick body verbatim aside
 * from input/output shape.
 *
 * @param totalMultiplier  G.totalMultiplier at call time — read only for the
 *                         t-chal 1 acceleratorEffect exponent
 *                         (`acceleratorPower ^ (totalAccelerator + totalMultiplier)`).
 *                         Separate from `input` because it's set by updateAllMultiplier
 *                         in the legacy ordering and we want the parity test to
 *                         exercise both with/without that completed.
 */
export function updateAllTick (
  input: UpdateAllTickInput,
  totalMultiplier: number
): UpdateAllTickResult {
  let a = 0

  const totalAcceleratorInit = input.acceleratorBought
  const costDivisor = 1

  if (input.upgrade8 !== 0) {
    a += Math.floor(input.multiplierBought / 7)
  }
  if (input.upgrade21 !== 0) {
    a += 5
  }
  if (input.upgrade22 !== 0) {
    a += 4
  }
  if (input.upgrade23 !== 0) {
    a += 3
  }
  if (input.upgrade24 !== 0) {
    a += 2
  }
  if (input.upgrade25 !== 0) {
    a += 1
  }
  if (input.upgrade32 !== 0) {
    a += Math.min(500, Math.floor(Decimal.log(input.prestigePoints.add(1), 1e25)))
  }
  if (input.upgrade45 !== 0) {
    a += Math.min(2500, Math.floor(Decimal.log(input.transcendShards.add(1), 10)))
  }
  a += input.acceleratorsAchievement

  const ecc2tr = CalcECC('transcend', input.c2Completions)
  const ecc7r = CalcECC('reincarnation', input.c7Completions)

  a += 5 * ecc2tr
  const freeUpgradeAccelerator = a

  a += input.totalAcceleratorBoost
    * (5
      + 2 * input.research18
      + 2 * input.research19
      + 3 * input.research20
      + input.acceleratorCubeBlessing)

  if (input.prestigeUnlocked) {
    a *= input.multiplicativeAcceleratorsRune
  }

  a *= input.acceleratorMultiplier
  a = Math.pow(
    a,
    Math.min(1, (1 + input.platonicUpgrade6 / 30) * input.viscosityPower)
  )
  a += input.hepteractAccelerators
  a *= input.challenge15RewardAccelerator
  a *= input.hepteractAcceleratorMult
  a = Math.floor(Math.min(1e100, a))

  if (input.viscosityCorruptionLevel >= 15) {
    a = Math.pow(a, 0.2)
  }
  if (input.viscosityCorruptionLevel >= 16) {
    a = 1
  }

  const freeAccelerator = a
  const totalAccelerator = totalAcceleratorInit + freeAccelerator

  const tuSevenMulti = input.upgrade46 > 0.5 ? 1.05 : 1

  let acceleratorPower = Math.pow(
    1.1
      + input.acceleratorPowerRune
      + 1 / 400 * ecc2tr
      + input.acceleratorPowerAchievement
      + tuSevenMulti
        * (input.totalAcceleratorBoost / 100)
        * (1 + ecc2tr / 20),
    1 + 0.04 * ecc7r
  )

  // No-MA and Sadistic challenges overwrite the transcension overrides
  if (
    input.reincarnationChallenge !== 7
    && input.reincarnationChallenge !== 10
  ) {
    if (input.transcensionChallenge === 1) {
      acceleratorPower *= 25 / (50 + input.c1Completions)
      acceleratorPower += 0.55
      acceleratorPower = Math.max(1, acceleratorPower)
    }
    if (input.transcensionChallenge === 2) {
      acceleratorPower = 1
    }
    if (input.transcensionChallenge === 3) {
      acceleratorPower = 1 + acceleratorPower / 2
    }
  }
  acceleratorPower = Math.min(1e300, acceleratorPower)
  if (input.reincarnationChallenge === 7) {
    acceleratorPower = 1
  }
  if (input.reincarnationChallenge === 10) {
    acceleratorPower = 1
  }

  let acceleratorEffect: Decimal
  if (input.transcensionChallenge !== 1) {
    acceleratorEffect = Decimal.pow(acceleratorPower, totalAccelerator)
  } else {
    acceleratorEffect = Decimal.pow(acceleratorPower, totalAccelerator + totalMultiplier)
  }
  const acceleratorEffectDisplay = new Decimal(acceleratorPower * 100 - 100)
  if (input.reincarnationChallenge === 10) {
    acceleratorEffect = new Decimal(1)
  }

  let generatorPower = new Decimal(1)
  if (input.upgrade11 > 0.5 && input.reincarnationChallenge !== 7) {
    generatorPower = Decimal.pow(1.02, totalAccelerator)
  }

  return {
    totalAccelerator,
    costDivisor,
    freeUpgradeAccelerator,
    freeAccelerator,
    tuSevenMulti,
    acceleratorPower,
    acceleratorEffect,
    acceleratorEffectDisplay,
    generatorPower
  }
}

File: crates/synergismforkd_logic/src/mechanics/update_all_tick.rs
//! Per-tick accelerator-state aggregator.
//!
//! Verbatim port of
//! `legacy/core_split/packages/logic/src/mechanics/updateAllTick.ts`.
//! Computes the full accelerator stack for the current tick.
//!
//! Takes a `&GameState` (for direct `player.*` reads) plus an
//! [`UpdateAllTickPre`] bundle (for cross-mechanic pre-computed values
//! that aren't in any slice) and `total_multiplier` (the output of
//! `update_all_multiplier` earlier this tick — used only in the
//! `transcensionChallenge == 1` branch of `accelerator_effect`).

use crate::mechanics::challenges::{calc_ecc, ChallengeType};
use crate::state::{GameState, VISCOSITY_INDEX};
use synergismforkd_bignum::Decimal;

/// Cross-mechanic pre-computed values the aggregator needs that don't
/// live in any state slice.
#[derive(Debug, Clone, Copy)]
pub struct UpdateAllTickPre {
    /// `+getAchievementReward('accelerators')`.
    pub accelerators_achievement: f64,
    /// `+getAchievementReward('acceleratorPower')`.
    pub accelerator_power_achievement: f64,
    /// `getRuneEffects('speed', 'multiplicativeAccelerators')`.
    pub multiplicative_accelerators_rune: f64,
    /// `getRuneEffects('speed', 'acceleratorPower')`.
    pub accelerator_power_rune: f64,
    /// `calculateAcceleratorCubeBlessing()`.
    pub accelerator_cube_blessing: f64,
    /// `getHepteractEffects('accelerator').accelerators`.
    pub hepteract_accelerators: f64,
    /// `getHepteractEffects('accelerator').acceleratorMultiplier`.
    pub hepteract_accelerator_mult: f64,
    /// `G.totalAcceleratorBoost`.
    pub total_accelerator_boost: f64,
    /// `G.acceleratorMultiplier`.
    pub accelerator_multiplier: f64,
    /// `G.viscosityPower[player.corruptions.used.viscosity]`.
    pub viscosity_power: f64,
    /// `G.challenge15Rewards.accelerator.value`.
    pub challenge_15_reward_accelerator: f64,
}

impl Default for UpdateAllTickPre {
    /// Identity values — multiplicative effects collapse to `1`, additive
    /// effects to `0`.
    fn default() -> Self {
        Self {
            accelerators_achievement: 0.0,
            accelerator_power_achievement: 0.0,
            multiplicative_accelerators_rune: 1.0,
            accelerator_power_rune: 0.0,
            accelerator_cube_blessing: 0.0,
            hepteract_accelerators: 0.0,
            hepteract_accelerator_mult: 1.0,
            total_accelerator_boost: 0.0,
            accelerator_multiplier: 1.0,
            viscosity_power: 1.0,
            challenge_15_reward_accelerator: 1.0,
        }
    }
}

/// Result of [`update_all_tick`].
#[derive(Debug, Clone, Copy)]
pub struct UpdateAllTickResult {
    /// `G.totalAccelerator`.
    pub total_accelerator: f64,
    /// `G.costDivisor` — always `1`.
    pub cost_divisor: f64,
    /// `G.freeUpgradeAccelerator`.
    pub free_upgrade_accelerator: f64,
    /// `G.freeAccelerator`.
    pub free_accelerator: f64,
    /// `G.tuSevenMulti`.
    pub tu_seven_multi: f64,
    /// `G.acceleratorPower`.
    pub accelerator_power: f64,
    /// `G.acceleratorEffect`.
    pub accelerator_effect: Decimal,
    /// `G.acceleratorEffectDisplay`.
    pub accelerator_effect_display: Decimal,
    /// `G.generatorPower`.
    pub generator_power: Decimal,
}

/// Per-tick accelerator-state aggregator. `total_multiplier` is the
/// post-`update_all_multiplier` value, used only in the
/// `transcensionChallenge == 1` branch of `acceleratorEffect`.
#[must_use]
#[allow(clippy::too_many_lines)]
pub fn update_all_tick(
    state: &GameState,
    pre: &UpdateAllTickPre,
    total_multiplier: f64,
) -> UpdateAllTickResult {
    let upgrade = |i: usize| f64::from(state.upgrades.upgrades[i]);
    let research = |i: usize| state.researches.researches[i];
    let c_completions = |i: usize| state.challenges.challenge_completions[i];

    let accelerator_bought = state.accelerator.accelerator_bought;
    let multiplier_bought = state.multiplier.multiplier_bought;
    let prestige_points = state.upgrades.prestige_points;
    let transcend_shards = state.reset_counters.transcend_shards;
    let transcension_challenge = state.challenges.current_transcension_challenge;
    let reincarnation_challenge = state.challenges.current_reincarnation_challenge;
    let platonic_upgrade_6 = state.cube_upgrade_levels.platonic_upgrades[6];
    let prestige_unlocked = state.reset_counters.prestige_unlocked;
    let viscosity_corruption_level = state.corruptions.used.levels[VISCOSITY_INDEX];

    let mut a = 0.0_f64;

    let total_accelerator_init = accelerator_bought;
    let cost_divisor = 1.0_f64;

    if upgrade(8) != 0.0 {
        a += (multiplier_bought / 7.0).floor();
    }
    if upgrade(21) != 0.0 {
        a += 5.0;
    }
    if upgrade(22) != 0.0 {
        a += 4.0;
    }
    if upgrade(23) != 0.0 {
        a += 3.0;
    }
    if upgrade(24) != 0.0 {
        a += 2.0;
    }
    if upgrade(25) != 0.0 {
        a += 1.0;
    }
    if upgrade(32) != 0.0 {
        let log_val = (prestige_points + Decimal::one())
            .log(Decimal::from_finite(1e25))
            .to_number()
            .floor();
        a += 500.0_f64.min(log_val);
    }
    if upgrade(45) != 0.0 {
        let log_val = (transcend_shards + Decimal::one())
            .log10()
            .to_number()
            .floor();
        a += 2500.0_f64.min(log_val);
    }
    a += pre.accelerators_achievement;

    let ecc2tr = calc_ecc(ChallengeType::Transcend, c_completions(2));
    let ecc7r = calc_ecc(ChallengeType::Reincarnation, c_completions(7));

    a += 5.0 * ecc2tr;
    let free_upgrade_accelerator = a;

    a += pre.total_accelerator_boost
        * (5.0
            + 2.0 * research(18)
            + 2.0 * research(19)
            + 3.0 * research(20)
            + pre.accelerator_cube_blessing);

    if prestige_unlocked {
        a *= pre.multiplicative_accelerators_rune;
    }

    a *= pre.accelerator_multiplier;
    a = a.powf(1.0_f64.min((1.0 + platonic_upgrade_6 / 30.0) * pre.viscosity_power));
    a += pre.hepteract_accelerators;
    a *= pre.challenge_15_reward_accelerator;
    a *= pre.hepteract_accelerator_mult;
    a = 1e100_f64.min(a).floor();

    if viscosity_corruption_level >= 15 {
        a = a.powf(0.2);
    }
    if viscosity_corruption_level >= 16 {
        a = 1.0;
    }

    let free_accelerator = a;
    let total_accelerator = total_accelerator_init + free_accelerator;

    let tu_seven_multi = if upgrade(46) > 0.5 { 1.05 } else { 1.0 };

    let mut accelerator_power = (1.1
        + pre.accelerator_power_rune
        + 1.0 / 400.0 * ecc2tr
        + pre.accelerator_power_achievement
        + tu_seven_multi * (pre.total_accelerator_boost / 100.0) * (1.0 + ecc2tr / 20.0))
        .powf(1.0 + 0.04 * ecc7r);

    if reincarnation_challenge != 7 && reincarnation_challenge != 10 {
        if transcension_challenge == 1 {
            accelerator_power *= 25.0 / (50.0 + c_completions(1));
            accelerator_power += 0.55;
            accelerator_power = 1.0_f64.max(accelerator_power);
        }
        if transcension_challenge == 2 {
            accelerator_power = 1.0;
        }
        if transcension_challenge == 3 {
            accelerator_power = 1.0 + accelerator_power / 2.0;
        }
    }
    accelerator_power = 1e300_f64.min(accelerator_power);
    if reincarnation_challenge == 7 {
        accelerator_power = 1.0;
    }
    if reincarnation_challenge == 10 {
        accelerator_power = 1.0;
    }

    let mut accelerator_effect = if transcension_challenge != 1 {
        Decimal::from_finite(accelerator_power).pow(Decimal::from_finite(total_accelerator))
    } else {
        Decimal::from_finite(accelerator_power)
            .pow(Decimal::from_finite(total_accelerator + total_multiplier))
    };
    let accelerator_effect_display = Decimal::from_finite(accelerator_power * 100.0 - 100.0);
    if reincarnation_challenge == 10 {
        accelerator_effect = Decimal::one();
    }

    let mut generator_power = Decimal::one();
    if upgrade(11) > 0.5 && reincarnation_challenge != 7 {
        generator_power = Decimal::from_finite(1.02).pow(Decimal::from_finite(total_accelerator));
    }

    UpdateAllTickResult {
        total_accelerator,
        cost_divisor,
        free_upgrade_accelerator,
        free_accelerator,
        tu_seven_multi,
        accelerator_power,
        accelerator_effect,
        accelerator_effect_display,
        generator_power,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_state_baseline() {
        let r = update_all_tick(&GameState::default(), &UpdateAllTickPre::default(), 0.0);
        // a = 0 → free_accelerator = 0.
        assert_eq!(r.free_accelerator, 0.0);
        assert_eq!(r.total_accelerator, 0.0);
        assert_eq!(r.tu_seven_multi, 1.0);
        // accelerator_power = (1.1 + 0 + …).powf(1) = 1.1.
        assert!((r.accelerator_power - 1.1).abs() < 1e-12);
    }

    #[test]
    fn upgrade_21_adds_5_to_a() {
        let mut state = GameState::default();
        state.upgrades.upgrades[21] = 1;
        let r = update_all_tick(&state, &UpdateAllTickPre::default(), 0.0);
        // a=5, then a^min(1, 1*1)=5; floor → 5.
        assert_eq!(r.free_accelerator, 5.0);
    }

    #[test]
    fn viscosity_16_zeros_to_1() {
        let mut state = GameState::default();
        state.corruptions.used.levels[VISCOSITY_INDEX] = 16;
        state.upgrades.upgrades[21] = 1;
        let r = update_all_tick(&state, &UpdateAllTickPre::default(), 0.0);
        assert_eq!(r.free_accelerator, 1.0);
    }

    #[test]
    fn reincarnation_chal_10_zeros_effect() {
        let mut state = GameState::default();
        state.challenges.current_reincarnation_challenge = 10;
        let r = update_all_tick(&state, &UpdateAllTickPre::default(), 0.0);
        assert_eq!(r.accelerator_effect.to_number(), 1.0);
        assert_eq!(r.accelerator_power, 1.0);
    }

    #[test]
    fn upgrade_46_sets_tu_seven_multi() {
        let mut state = GameState::default();
        state.upgrades.upgrades[46] = 1;
        let r = update_all_tick(&state, &UpdateAllTickPre::default(), 0.0);
        assert_eq!(r.tu_seven_multi, 1.05);
    }

    #[test]
    fn generator_power_active_with_upgrade_11_outside_chal_7() {
        let mut state = GameState::default();
        state.upgrades.upgrades[11] = 1;
        state.accelerator.accelerator_bought = 100.0;
        let r = update_all_tick(&state, &UpdateAllTickPre::default(), 0.0);
        // 1.02^100 ≈ 7.24.
        let expected = 1.02_f64.powi(100);
        assert!((r.generator_power.to_number() - expected).abs() < 1e-6);
    }
}
